senders/solomon/solomon.go

```diff
@@ -226,38 +240,123 @@ func (s *solomonClient) Send(task tasks.DataType, timestamp uint64) error {
 		}
 		logger.Debugf("%s Data to POST: %s", s.id, string(push))
 
-		req, err := http.NewRequest("POST", s.api, bytes.NewBuffer(push))
+		logger.Debugf("%s Sending work to JobQueue", s.id)
+		JobQueue <- Job{PushData: push, SolCli: s}
+
+	}
+
+	return nil
+}
+
+func NewSolomonClient(cfg *SolomonCfg, id string) (ss SolomonSender, err error) {
+	ss = &solomonClient{
+		id:                 id,
+		api:                cfg.Api,
+		cluster:            cfg.Cluster,
+		project:            cfg.Project,
+		fields:             cfg.Fields,
+		connection_timeout: cfg.Connection_timeout,
+		rw_timeout:         cfg.Rw_timeout,
+	}
+
+	return
+}
+
+func (w Worker) Start(d *Dispatcher) {
+	for {
+		// add current worker to the WorkerPool on first iteration
+		// after finishing select statement worker will add
+		// itself back to the pool again
+		d.WorkerPool <- w.JobChannel
+		select {
+		// wait for a job in workers JobChannel
+		case job := <-w.JobChannel:
+			logger.Debugf("%s worker %d received job", job.SolCli.id, w.Id)
+			if err := w.SendToSolomon(job); err != nil {
+				logger.Errf("%s", err)
+
+			}
+		}
+	}
+}
+
+func (w Worker) SendToSolomon(job Job) error {
+	for i := 1; i <= w.Retry; i++ {
+		SolomonHTTPClient := httpclient.NewClientWithTimeout(
+			time.Millisecond*(time.Duration(job.SolCli.connection_timeout)),
+			time.Millisecond*(time.Duration(job.SolCli.rw_timeout)))
+		req, err := http.NewRequest("POST", job.SolCli.api, bytes.NewBuffer(job.PushData))
 		if err != nil {
-			return fmt.Errorf("%s %s", s.id, err)
+			return fmt.Errorf("%s %s", job.SolCli.id, err)
 		}
 		req.Header.Add("Content-Type", "application/json")
-
+		logger.Debugf("%s Attempting to send. Worker %d. Attempt %d", job.SolCli.id, w.Id, i)
 		resp, err := SolomonHTTPClient.Do(req)
+
 		if err != nil {
-			return fmt.Errorf("%s %s", s.id, err)
+			return fmt.Errorf("%s %s", job.SolCli.id, err)
 		}
 		defer resp.Body.Close()
 
+		if resp.StatusCode == http.StatusRequestTimeout {
+			if i == (w.Retry) {
+				return fmt.Errorf("%s failed to send after %d attemps. Worker %d. Dropping", job.SolCli.id, i, w.Id)
+
+			}
+			logger.Debugf("%s timed out. Worker %s. Retrying.", job.SolCli.id, w.Id)
+			time.Sleep(time.Second * 1)
+			continue
+
+		}
+
 		if resp.StatusCode != http.StatusOK {
 			b, err := ioutil.ReadAll(resp.Body)
 			if err != nil {
-				return fmt.Errorf("%s bad response code '%d': %s", s.id, resp.StatusCode, resp.Status)
+				return fmt.Errorf("%s bad response code '%d': %s", job.SolCli.id, resp.StatusCode, resp.Status)
 			}
-			return fmt.Errorf("%s bad response code '%d' '%s': %s", s.id, resp.StatusCode, resp.Status, b)
+			return fmt.Errorf("%s bad response code '%d' '%s': %s", job.SolCli.id, resp.StatusCode, resp.Status, b)
+		} else {
```

Comment:
redundant `else`
if branch has return statement, so there is no sense in havind `else`